Fix: Zeebe gateway fails to stream out activated jobs by not respecting the max message size

berkaycanbc opened the pull request:
## Description

<!-- Please explain the changes you made here. -->

There were three options to solve original problem:
1) When mapping the broker response to the client response, do bookkeeping
on the response size to ensure the maxMessageSize.
2) Double the max inbound message size in the Zeebe Java Client by default.
3) Bring the WARNING: Stream Error exception into the control flow of the Gateway
so that the Gateway can react to it and potentially make the jobs re-activatable again.

I used mixed of option 1 and 3 for better UX. With option 2, it is still possible for
client users to override default maxInboundMessage which might result in the same issue.

In the ResponseMapper, we check if the gRPC response size of the activated job will
exceed the allowed maxMessageSize in the gateway. Afterward, we re-activate exceeding
jobs in RoundRobinActivateJobsHandler in a similar way that we do for failed jobs.

Using `getSerializedSize()` method shouldn't have any performance impact here since
the result of it will be cached to be used while actually returning the response.
See: https://protobuf.dev/reference/java/api-docs/com/google/protobuf/AbstractMessage.html#getSerializedSize--

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #12778 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

korthout commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/network/LargeMessageSizeTest.java:
```diff
@@ -134,4 +151,64 @@ public void shouldCompleteJobWithLargeVariables() {
+              .send()
+              .join();
+
+      RecordingExporter.processInstanceRecords()
+          .withProcessInstanceKey(event.getProcessInstanceKey())
+          .withElementType(BpmnElementType.PROCESS)
+          .await();
```
🔧 I think we don't have to wait for the command to be appended to the log in the loop because sending the 5 create instance commands and processing them asynchronously should be fine. I don't think it affects the rest of this test case.

berkaycanbc replied:
@korthout is it because the instances themselves are lightweight so will be created fast enough? If so, could you give me some examples where we should use `await()` method? I am asking to clarify for the next time I use it. 👍

berkaycanbc replied:
Definitely more clear, thanks! 🚀

korthout commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/network/LargeMessageSizeTest.java:
```diff
@@ -134,4 +151,64 @@ public void shouldCompleteJobWithLargeVariables() {
+    final JobWorkerBuilderStep3 builder =
+        CLIENT_RULE.getClient().newWorker().jobType("foo").handler(COMPLETING_JOB_HANDLER);
+
+    // then
+    try (final JobWorker ignored = builder.open()) {
+      Awaitility.await("until all jobs are completed")
+          .pollInterval(Duration.ofMillis(100))
+          .atMost(Duration.ofSeconds(5))
+          .untilAsserted(
+              () ->
+                  assertThat(
+                          RecordingExporter.jobRecords(JobIntent.COMPLETED)
+                              .withType("foo")
+                              .limit(numberOfJobsToActivate)
+                              .count())
+                      .isEqualTo(numberOfJobsToActivate));
+    }
+  }
```
👍 I like the test case!

korthout commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/network/LargeMessageSizeTest.java:
```diff
@@ -134,4 +151,64 @@ public void shouldCompleteJobWithLargeVariables() {
+          .withElementType(BpmnElementType.PROCESS)
+          .await();
+    }
+
+    // when
+    final JobWorkerBuilderStep3 builder =
+        CLIENT_RULE.getClient().newWorker().jobType("foo").handler(COMPLETING_JOB_HANDLER);
```
❌ Before opening the job worker, we must ensure that all 5 jobs exist. Otherwise, the job worker might activate jobs and receive less than 5 jobs.

berkaycanbc replied:
Yeah, makes sense! Updated.

korthout replied:
❌ This is still missing in both this test and the one above

korthout reviewed (changes requested):
Thanks @berkaycanbc this looks like a great addition!

⏸️ I haven't come as far as reviewing the implementation, but the test looks good and I like the main idea for the implementation.

❌ However, I thought I would already send out this review because there's a change required in the test case. Please have a look.

korthout commented on gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java:
```diff
@@ -276,22 +278,30 @@ public static SetVariablesResponse toSetVariablesResponse(
 
-  public static ActivateJobsResponse toActivateJobsResponse(
-      final long key, final JobBatchRecord brokerResponse) {
+  public static JobActivationResult toActivateJobsResponse(
```
🔧 I'd like to see some Javadoc on this method, to clarify why the result contains a response **and** jobs to defer, and why a maxMessageSize is needed.

🔧 When writing that, we may choose to rename this maxMessageSize to maxResponseSize, as this better clarifies the parameter in the context of the method.

💭 I also wonder whether jobsToDefer is a fitting name in this method. In the context of the method, something like jobsExceedingSize, jobsThatDontFitInTheResponse, or something along those lines, may fit better semantically.

berkaycanbc replied:
Good points, updated!

korthout commented on gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java:
```diff
@@ -276,22 +278,30 @@ public static SetVariablesResponse toSetVariablesResponse(
     final Iterator<LongValue> jobKeys = brokerResponse.jobKeys().iterator();
     final Iterator<JobRecord> jobs = brokerResponse.jobs().iterator();
 
+    int currentResponseSize = 0;
```
❌ I believe this is not entirely correct. I assume there is some metadata associated to an `ActivateJobsResponse`. Currently, only the size of the individual `ActivatedJob`s are taken into account.

You could try creating an empty `ActivateJobsResponse` object to see the serialized size, and then doing the same with one, two, and three added `ActivatedJob`s to see how it grows. This will give you the insight what the initial size should be.

berkaycanbc replied:
@korthout 


```java
final ActivateJobsResponse response =
        ActivateJobsResponse.newBuilder()
            .addJobs(ActivatedJob.newBuilder().build())
            .build();
System.out.println("response size: " + response.getSerializedSize()); // prints 2

final ActivatedJob activatedJob1 = ActivatedJob.newBuilder().build();
    System.out.println("activatedJob empty size: " + activatedJob1.getSerializedSize()); // prints 0
```

It only adds 2 bytes for each job added to ActivateJobsResponse and 0 for an empty activated job. I updated the code based on that, please take a look. 👀

berkaycanbc commented on gateway/src/main/java/io/camunda/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java:
```diff
@@ -27,6 +27,7 @@
 import java.util.Queue;
 import java.util.concurrent.ConcurrentHashMap;
 import org.slf4j.Logger;
+import org.springframework.util.unit.DataSize;
```
Very good point! 🚀 Updated.

berkaycanbc commented on gateway/src/main/java/io/camunda/zeebe/gateway/impl/job/RoundRobinActivateJobsHandler.java:
```diff
@@ -44,17 +46,22 @@ public final class RoundRobinActivateJobsHandler implements ActivateJobsHandler
+      final BrokerClient brokerClient, final DataSize maxMessageSize) {
     this.brokerClient = brokerClient;
     topologyManager = brokerClient.getTopologyManager();
+    this.maxMessageSize = (int) maxMessageSize.toBytes();
```
🤦‍♂️. Updated!

berkaycanbc commented:
Very useful feedback, thanks @korthout! ⭐ 

I applied your suggestions, please take a look. 👀

korthout commented on gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java:
```diff
@@ -276,22 +278,53 @@ public static SetVariablesResponse toSetVariablesResponse(
       final ActivatedJob activatedJob = toActivatedJob(jobKey.getValue(), job);
 
-      responseBuilder.addJobs(activatedJob);
+      // +2 is for the metadata size that is added to ActivateJobsResponse when a new ActivatedJob
+      // is added to it
+      currentResponseSize += activatedJob.getSerializedSize() + 2;
```
🤔 This looks hard to maintain. If we ever add something to `ActivatedJob` or `ActivateJobsResponse` then this is no longer correct.

🔧 How about adding a post-condition check at the end of this method?
```java
if (currentResponseSize != response.getSerializedSize()) {
 // we've calculated the current response size incorrectly
}
```

berkaycanbc replied:
@korthout If that is the case, how are we going to handle it? Are we going to put all jobs to `sizeExceedingJobs`? Or take the last job from the responseBuilder jobs list and put it into `sizeExceedingJobs`?

korthout replied:
When that happens, it means we have a bug. We can:
- log an error so users can report the bug
- throw an exception and reject the command
- or somehow add a test case that fails when the size is calculated wrongly (seems difficult)

I don't see a need to complicate the job selection logic for this. If we calculate the size wrongly **and** the response exceeds the size limit, then this will result in an error again.

My personal preference is simply logging an error, hoping that a user reports it.

berkaycanbc replied:
@korthout The if check suggestion you provided will not do the work. Because current response size will always be different than the `response.getSerializedSize()` (e.g. serialized size will be +2).

berkaycanbc replied:
I would either keep `+2` or not add it at all. Even if a new field is added to the ActivateJobsResponse and we faced with exceeding size failures, it will be easier to add a fix for that. WDYT? @korthout

korthout replied:
>The if check suggestion you provided will not do the work. Because current response size will be different than the `response.getSerializedSize()` even if we add jobs to exceeding list.

I don't understand what you mean. 

❓ In my mind, the goal is to not exceed the max message size with the response because that may break things. We added a way to calculate the response size while adding jobs to the response. That calculation is 0 bytes for the response + for each job: 'size of the job + 2 bytes for metadata'. Just before returning the response, we can also request the actual size of the response with `response.getSerializedSize()` which should be equivalent to the manually calculated size. Or is that not the case? Am I missing something?

berkaycanbc replied:
Ohh :

berkaycanbc replied:
Ohh 😯. You mean keeping the `+2` and if we forget to update the value on field updates, we log the error. Then, we expected to see these errors in log and fix the issue. Is that correct? @korthout 

I thought it is hard to maintain +2, so we should remove it.

korthout replied:
> You mean keeping the +2 and if we forget to update the value on field updates, we log the error. Then, we expected to see these errors in log and fix the issue. Is that correct?

Yes, that's correct. I would keep that logic while it's correct, but add a check on top that tells us that the logic is correct. If things ever change, we might not have a failing test for this, but at least we have an ERROR log message that users could report.

berkaycanbc commented:
@korthout I found that the calculation of serialised size also depends of the length of returned response field. Looks like making the metadata size addition static will not be correct.

See related google protobuf implementations:
https://github.com/protocolbuffers/protobuf/blob/main/java/core/src/main/java/com/google/protobuf/CodedOutputStream.java#L607
https://github.com/protocolbuffers/protobuf/blob/main/java/core/src/main/java/com/google/protobuf/CodedOutputStream.java#L700


I will give a try to the re-building response until the actual response size is under max message size. 👍

berkaycanbc commented:
@korthout I added another if check to the end of response mapper in case gRPC response exceeds max message size after building it. (metadata will be added)

I also added a test case in another commit by configuring it with some magic numbers. I try to add as much as explanation as possible to make the test case clear.

Please take a look. 👀

korthout commented on gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java:
```diff
@@ -301,30 +301,42 @@ public static SetVariablesResponse toSetVariablesResponse(
+    // is still exceeding the maximum response size, we remove the last added job from the response
+    // and add it to the list of jobs to be reactivated.
+    // We do this until the response size is below the maximum response size.
+    if (response.getSerializedSize() > maxResponseSize) {
+      while (!responseJobs.isEmpty() && response.getSerializedSize() > maxResponseSize) {
```
🔧 You could remove the if statement, as the while already covers it on its own.

berkaycanbc replied:
Ahh, right, I missed it after the refactor. 👍

korthout commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/network/LargeMessageSizeTest.java:
```diff
@@ -134,4 +150,139 @@ public void shouldCompleteJobWithLargeVariables() {
+    final JobWorkerBuilderStep3 builder =
+        CLIENT_RULE.getClient().newWorker().jobType("foo").handler(COMPLETING_JOB_HANDLER);
+
+    RecordingExporter.jobRecords(JobIntent.CREATED)
+        .withType("foo")
+        .limit(numberOfJobsToActivate)
+        .await();
```
💭 I think this does something else than you expect. While you limit the stream to 5 records, it only throws an exception when the stream has no entries after 5 seconds. It would wait 5 seconds and then continue if there is only a single job record.

berkaycanbc replied:
Just learned that it works that way, thanks! 🚀  Now, I assert the job count. 👍

korthout reviewed (changes requested):
Nice @berkaycanbc 🙌 

👍 I like the new approach with a fast happy path and an easy to understand edge case.

❌ The test cases could be flaky, please have a look

berkaycanbc commented:
@korthout I sent updates. Please take a look. 👀

korthout reviewed (changes requested):
Thanks @berkaycanbc 

❌ There are still some things to fix:
- unused local variables on lines 179 and 249
- unnecessary blocking `join` on lines 187 and 257
- slow tests due to missing `limit` on lines 194 and 264
- test speed improvement possible by awaiting for all 5 jobs created before opening job worker (avoids multiple activation roundtrips)

berkaycanbc commented:
@korthout I sent updates, please take a look. 👀 

> test speed improvement possible by awaiting for all 5 jobs created before opening job worker (avoids multiple activation roundtrips)

I am not sure which change is need for this one tho. I expect asserting job creation should do the work. Please feel free to add a code suggest if you think there is additional change needed for that. There is a support case waiting for this issue to be merged, so that we can speed this up to unblock customer. 👍

korthout reviewed (approved):
Thanks for the last changes @berkaycanbc ❤️ 

👍 LGTM

korthout commented:
@berkaycanbc As you'll want to backport this PR, please rebase it on top of main rather than merging main in with a merge commit.

If you struggle with resolving conflicts multiple times, I recommend enabling [`git rerere`](https://git-scm.com/book/en/v2/Git-Tools-Rerere).

berkaycanbc commented:
> @berkaycanbc As you'll want to backport this PR, please rebase it on top of main rather than merging main in with a merge commit.
> 
> If you struggle with resolving conflicts multiple times, I recommend enabling [`git rerere`](https://git-scm.com/book/en/v2/Git-Tools-Rerere).

Thanks @korthout, rebased onto main. 👍

berkaycanbc commented:
Code scanning has a false positive, triggering merge.

bors merge

berkaycanbc commented:
> This PR was included in a batch that successfully built, but then failed to merge into main. It will not be retried.
> 
> Additional information:
> 
> ```json
> {"message":"Changes must be made through the merge queue You're not authorized to push to this branch. Visit https://docs.github.com/articles/about-protected-branches/ for more information.","documentation_url":"https://docs.github.com/articles/about-protected-branches"}
> ```

Any ideas @korthout ?

backport-action commented:
Backport failed for `stable/8.2`, because it was unable to cherry-pick the commit(s).

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin stable/8.2
git worktree add -d .worktree/backport-15784-to-stable/8.2 origin/stable/8.2
cd .worktree/backport-15784-to-stable/8.2
git switch --create backport-15784-to-stable/8.2
git cherry-pick -x a6a6b95f7166289bca17e5502f5eb2fa38018f7d 58c0e20183ebc2faf03f38dd1109b21118a1f586 b2ad953cee58b9675b5f64a901f17a9ff412d937 d5e3debb1e13915e2885592a58009e2feb2179ac
```

backport-action commented:
Backport failed for `stable/8.3`, because it was unable to cherry-pick the commit(s).

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin stable/8.3
git worktree add -d .worktree/backport-15784-to-stable/8.3 origin/stable/8.3
cd .worktree/backport-15784-to-stable/8.3
git switch --create backport-15784-to-stable/8.3
git cherry-pick -x a6a6b95f7166289bca17e5502f5eb2fa38018f7d 58c0e20183ebc2faf03f38dd1109b21118a1f586 b2ad953cee58b9675b5f64a901f17a9ff412d937 d5e3debb1e13915e2885592a58009e2feb2179ac
```

backport-action commented:
Backport failed for `stable/8.4`, because it was unable to cherry-pick the commit(s).

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin stable/8.4
git worktree add -d .worktree/backport-15784-to-stable/8.4 origin/stable/8.4
cd .worktree/backport-15784-to-stable/8.4
git switch --create backport-15784-to-stable/8.4
git cherry-pick -x a6a6b95f7166289bca17e5502f5eb2fa38018f7d 58c0e20183ebc2faf03f38dd1109b21118a1f586 b2ad953cee58b9675b5f64a901f17a9ff412d937 d5e3debb1e13915e2885592a58009e2feb2179ac
```
